### PR TITLE
Roles updated

### PR DESCRIPTION
**Pull request by differentreality**
#583

### CODE REVIEW

**Comment by differentreality** [comment]
This is also ready for review

screenshot-osem - chromium
screenshot-osem - chromium-1

File: app/controllers/admin/conference_controller.rb [hunk 1]
```diff
@@ -169,29 +173,12 @@ def edit
...
-      render 'roles', formats: [:js]
-    end
-
-    def remove_user
-      @selection = params[:role]
-      @user.revoke @selection, @conference
-
-      @role_users = get_users(@selection)
-      render 'roles', formats: [:js]
+    def create_roles
```
Comment (ChrisBr) [inline app/controllers/admin/conference_controller.rb]: Hm, I don't like this in the controller ... 
What about do this in a [callback](http://api.rubyonrails.org/classes/ActiveRecord/Callbacks.html)?

Comment (ChrisBr) [inline app/controllers/admin/conference_controller.rb]: Of course, you would need to rewrite your conditional in line ``72`` also...

File: app/controllers/admin/conference_controller.rb [hunk 2]
```diff
@@ -67,11 +66,16 @@ def create
...
       if @conference.valid?
         @conference.save
+        flash[:notice] = 'Conference was successfully created.'
+
         # user that creates the conference becomes organizer of that conference
-        current_user.add_role :organizer, @conference
-        redirect_to(admin_conference_path(id: @conference.short_title),
-                    notice: 'Conference was successfully created.')
+        unless create_roles && (current_user.add_role :organizer, @conference)
+         flash[:alert] = 'Could not create roles properly!'
```
Comment (ChrisBr) [inline app/controllers/admin/conference_controller.rb]: Do we than an alert and a notice flash?
I don't like that ...

File: app/controllers/admin/conference_controller.rb [hunk 3]
```diff
@@ -67,11 +66,16 @@ def create
 
       if @conference.valid?
```
Comment (ChrisBr) [inline app/controllers/admin/conference_controller.rb]: If you touch this action anyway, what about refactor this to 
``
@conference.save
``
instead of
``
if @conference.valid?
  @conference.save
  ...
``

it's done twice ...

File: app/controllers/admin/roles_controller.rb [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+    end
+
+    def edit
+      @users = @role.users
+    end
+
+    def update
+      role_name = @role.name
+
+      if @role.name == 'organizer' && params[:role][:user_ids].reject(&:empty?).blank?
```
Comment (ChrisBr) [inline app/controllers/admin/roles_controller.rb]: Hm, looks like a validation?

File: app/controllers/admin/roles_controller.rb [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+        redirect_to admin_conference_role_path(@conference.short_title, @role.name)
+      else
+        flash[:error] = 'Could not update role! ' + @role.errors.full_messages.to_sentence
+        @role.name = role_name
+        render :edit
+      end
+    end
+
+    def toggle_user
+      if params[:user] && (user_email = params[:user][:email]) && user_email.present?
```
Comment (ChrisBr) [inline app/controllers/admin/roles_controller.rb]: Please move 
``
user_email = params[:user][:email]
``
out of the conditional! This is really hard to understand ...

File: app/controllers/admin/roles_controller.rb [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+        user = User.find_by(email: user_email)
+        if user
+          state = params[:user][:state]
+          if state == 'false'
+            @role.users.delete(user)
+            flash[:notice] = "User #{user.email} removed successfully"
+          else
+            @role.users << user unless @role.users.include? user
```
Comment (ChrisBr) [inline app/controllers/admin/roles_controller.rb]: This looks like a validation
``unless @role.users.include? user``

File: app/controllers/admin/roles_controller.rb [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+        flash[:error] = 'You must enter a search word!'
+      end
+      redirect_to admin_conference_role_path(@conference.short_title, @role.name)
+    end
+
+    protected
+
+    def set_selection
+      # Set 'organizer' as default role, when there is no other selection
+      params[:id] ? (@selection = params[:id].parameterize.underscore) : (@selection = 'organizer')
```
Comment (ChrisBr) [inline app/controllers/admin/roles_controller.rb]: This is more readable
``
@selection = params[:id] ? params[:id].parameterize.underscore : 'organizer'
``

File: app/models/role.rb [hunk 8]
```diff
@@ -1,9 +1,25 @@
 class Role < ActiveRecord::Base
-  has_and_belongs_to_many :users
   belongs_to :resource, polymorphic: true
+  has_and_belongs_to_many :users
 
   scopify
 
-  LABELS = ['Attendee', 'Volunteer', 'Speaker', 'Sponsor', 'Press', 'Keynote Speaker']
```
Comment (ChrisBr) [inline app/models/role.rb]: Why did you remove it?

**Comment by ChrisBr** [comment]
And you'd need to rebase!
Otherwise good work @differentreality

File: app/models/role.rb (hunk 8 above)
Comment (differentreality) [inline app/models/role.rb]: Roles will only be about actionable roles and we will not have labels mixed with roles.

File: app/controllers/admin/roles_controller.rb (hunk 6 above)
Comment (differentreality) [inline app/controllers/admin/roles_controller.rb]: Is it? I am not seeing it.
I just don't want to excecute code that does not need to be executed.

Re-writing it another (simpler?) way anyway!

File: app/controllers/admin/conference_controller.rb (hunk 1 above)
Comment (differentreality) [inline app/controllers/admin/conference_controller.rb]: Sth goes terribly wrong when I put it in after_save.
I will further investigate. Ideas welcome.

Comment (hennevogel) [inline app/controllers/admin/conference_controller.rb]: Can you tell us what goes wrong?

Comment (differentreality) [inline app/controllers/admin/conference_controller.rb]: First of all the fact that I don't know what this is doing here again... It is moved in models/conference.rb (see below)

Comment (hennevogel) [inline app/controllers/admin/conference_controller.rb]: Heh, so remove it here please :-)

**Comment by hennevogel** [comment]
Alright, let's get this in!
